vortex-ffi/src/sink.rs

```diff
@@ -0,0 +1,92 @@
+use std::ffi::{CStr, c_char};
+use std::ptr;
+
+use mpsc::Sender;
+use tokio::fs::File;
+use tokio::sync::mpsc;
+use tokio::task::JoinHandle;
+use tokio_stream::wrappers::ReceiverStream;
+use vortex::ArrayRef;
+use vortex::dtype::DType;
+use vortex::error::{VortexExpect, VortexResult, vortex_err};
+use vortex::file::VortexWriteOptions;
+use vortex::stream::ArrayStreamAdapter;
+
+use crate::RUNTIME;
+use crate::array::vx_array;
+use crate::error::{try_or, vx_error};
+
+#[allow(non_camel_case_types)]
+/// The `sink` interface is used to collect array chunks and place them into a resource
+/// (e.g. an array stream or file (`vx_array_sink_open_file`)).
+pub struct vx_array_sink {
+    sink: Sender<VortexResult<ArrayRef>>,
+    writer: JoinHandle<VortexResult<File>>,
+}
+
+/// Opens a writable array stream, where sink is used to push values into the stream.
+/// To close the stream close the sink with `vx_array_sink_close`.
+#[unsafe(no_mangle)]
+pub unsafe extern "C-unwind" fn vx_array_sink_open_file(
+    path: *const c_char,
+    dtype: *const DType,
+    error: *mut *mut vx_error,
+) -> *mut vx_array_sink {
+    try_or(error, ptr::null_mut(), || {
+        let path = unsafe { path.as_ref() }.vortex_expect("null path");
+        let path = unsafe { CStr::from_ptr(path) }
+            .to_string_lossy()
```

Comment:
Not clear that we should do this for file paths, vs returning an error. But we can change this as part of the FFI audit issue